[Resources] Add `cpus` in resource specification

WoosukKwon opened the pull request:
<!-- Describe the changes in this PR -->
Closes #387

This PR introduces the ability for users to specify resource requirements in terms of the number of vCPUs in the `Resources` class and the `resources` section in the SkyPilot yaml.

**Semantics**
1. If the instance type and accelerators are not specified, SkyPilot will use the `cpus` parameter to select an appropriate VM size from the cloud's **default** VM family (i.e., m6i for AWS, Dv5 for Azure, and n2-standard for GCP).

For example,
* `sky launch --cpus 8` or `Resources(cpus=8)`: SkyPilot will find the VMs that have exactly 8 vCPUs. Namely,
  * AWS: m6i.2xlarge
  * Azure: Standard_D8_v5
  * GCP: n2-standard-8

* `sky launch --cpus 12+` or `Resources(cpus='12+')`: SkyPilot will consider the VMs that have 12 or more vCPUs. Namely,
  * AWS: m6i.4xlarge
  * Azure: Standard_D16_v5
  * GCP: n2-standard-16

(Note: The selected three instance types have 16 vCPUs instead of 12 vCPUs, because the default instance families of the clouds do not include any VM with 12-15 vCPUs.)

2. If the accelerators are specified, the instance family can be inferred for each cloud, and SkyPilot will use the `cpus` parameter to choose the VM of the right size in the instance family. For example,
* `sky launch --gpus T4 --cpus 8+` or `Resources(accelerators='T4', cpus='8+')`: SkyPilot will choose VMs that have 1 T4 GPU along with 8 or more vCPUs. Namely,
  * AWS: g4dn.2xlarge'
  * Azure: Standard_NC8as_T4_v3
  * GCP: n1-highmem-8 (+ T4)

The same semantics applies to `cpunode`, `gpunode`, `tpunode`, and `spot launch`.


<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

concretevitamin commented on sky/clouds/service_catalog/aws_catalog.py:
```diff
@@ -94,6 +98,17 @@ def get_vcpus_from_instance_type(instance_type: str) -> Optional[float]:
 
+def get_default_instance_type(cpu: Optional[str] = None) -> Optional[str]:
+    if cpu is None:
+        cpu = '8'
```
Define as a constant on L25?

WoosukKwon replied:
Done.

concretevitamin reviewed (commented):
This is awesome @WoosukKwon! Had a quick scan and the approach makes sense to me. 

Quick question: are we deferring adding CPU as a scheduling constraint?

E.g., what happens if we launch a node with 8 CPUs, then keep `sky launch -c existing --cpu 4` for many times? Or, for `--cpu 16` do we correctly throw an error? 

(I think it's okay to defer, just checking.)

WoosukKwon commented:
@concretevitamin Thanks for the quick feedback. Yes, I'd like to defer it.

dongreenberg commented:
Love it!! Two questions:
1) I'm assuming this will work with mixing and matching inside the Resources() python API? e.g. I could do 
`Resources(accelerators='A100:1', cpus='8+')`?
2) Do all of these launch with the same OS an architecture?

WoosukKwon commented:
Hey @dongreenberg, thanks for your questions!

> 1. I'm assuming this will work with mixing and matching inside the Resources() python API? e.g. I could do
Resources(accelerators='A100:1', cpus='8+')?

Yes. It will work on the Python API in the same way.

> 2. Do all of the CPU-only boxes launch with the same OS and architecture?

They all have recent Intel x86-64 CPUs (i.e., either Ice Lake or Cascade Lake CPUs), and they all have 4 GB RAM per 1 vCPU. More specifically,

* AWS m6i series have Intel Ice Lake 8375C CPUs
* Azure Standard Dv5 series have Ice Lake 8370C CPUs
* GCP N2 series have Intel Ice Lake 8373C or Cascade Lake 6268CL CPUs

The OS might be different across the clouds and the VM images you select.

dongreenberg commented:
Amazing, that's perfect. You guys are the best!

concretevitamin commented on sky/resources.py:
```diff
@@ -45,12 +45,13 @@ class Resources:
     # 1. Increment the version. For backward compatibility.
     # 2. Change the __setstate__ method to handle the new fields.
     # 3. Modify the to_config method to handle the new fields.
```
This method name is out of date :)

WoosukKwon replied:
Fixed. PTAL.

concretevitamin commented on sky/resources.py:
```diff
@@ -45,12 +45,13 @@ class Resources:
         self,
         cloud: Optional[clouds.Cloud] = None,
         instance_type: Optional[str] = None,
+        cpu: Union[None, int, float, str] = None,
```
Should we update / add a TODO in less_demanding_than() / is_same_resources()?

concretevitamin replied:
reminder

WoosukKwon replied:
For `is_same_resources`, we don't need to do anything in this PR, because it's not reachable.

For ` less_demanding_than()`, it does not make sense to compare the user-specified `Resources` against the existing cluster's `Resources` in terms of `cpus`, because the latter one always has the `cpus` field set to None.

concretevitamin replied:
Thanks. For code gardening purpose, let's remove `is_same_resources` in this PR too?

WoosukKwon replied:
Removed.

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -180,7 +180,8 @@ def get_accelerators_from_instance_type(
     @classmethod
-    def get_default_instance_type(cls) -> str:
+    def get_default_instance_type(cls,
+                                  cpu: Optional[str] = None) -> Optional[str]:
```
Document why this would be None for some Cloud? And what happens in that case?

WoosukKwon replied:
If `cpu` is None, the return value will never be None. The return values can be None when `cpu` is set to an infeasible number (e.g., '123' or '1000+').

WoosukKwon replied:
Added the docstring. PTAL.

concretevitamin commented on sky/clouds/aws.py:
```diff
@@ -350,16 +349,21 @@ def _make(instance_list):
+            if default_instance_type is None:
+                return ([], fuzzy_candidate_list)
+            else:
+                return (_make([default_instance_type]), fuzzy_candidate_list)
```
Was confused a bit. Maybe replace `fuzzy_candidate_list` on L341, 364, 366 with []?

WoosukKwon replied:
Yes, they all are []. I thought it's a style convention to clarify the meaning, and mimicked the previous code. Changed them to [].

concretevitamin commented on sky/clouds/azure.py:
```diff
@@ -265,23 +265,30 @@ def _make(instance_list):
+            if default_instance_type is None:
+                return ([], fuzzy_candidate_list)
+            else:
+                return (_make([default_instance_type]), fuzzy_candidate_list)
```
ditto on replacing fuzzy_candidate_list?

WoosukKwon replied:
Changed to [].

concretevitamin commented on sky/clouds/service_catalog/__init__.py:
```diff
@@ -159,6 +159,12 @@ def get_vcpus_from_instance_type(instance_type: str,
 
+def get_default_instance_type(cpu: Optional[str] = None,
+                              clouds: CloudFilter = None) -> Optional[str]:
+    """Returns the default instance type for the given number of CPU."""
```
Shall we document the two formats?

WoosukKwon replied:
Added.

concretevitamin commented on sky/clouds/service_catalog/common.py:
```diff
@@ -313,6 +333,10 @@ def get_instance_type_for_accelerator_impl(
         return (None, fuzzy_candidate_list)
 
+    result = _filter_with_cpu(result, cpu)
+    if len(result) == 0:
+        return ([], [])
```
nit: duplicates L345?

WoosukKwon replied:
Good catch. Fixed.

concretevitamin commented on sky/clouds/service_catalog/common.py:
```diff
@@ -270,6 +270,25 @@ def get_vcpus_from_instance_type_impl(
+def _filter_with_cpu(df: pd.DataFrame, cpu: Optional[str]) -> pd.DataFrame:
+    if cpu is None:
+        return df
+    if cpu.endswith('+'):
+        return df[df['vCPUs'] >= float(cpu[:-1])]
+    else:
+        return df[df['vCPUs'] == float(cpu)]
```
nit: To be safer, maybe catch ValueError again?

WoosukKwon replied:
I think we shouldn't raise the error when the empty dataframe is returned, because there might exist other clouds that matches the requested number of vCPUs. If all clouds do not have a matching VM, the optimizer will then print the error msg.

concretevitamin replied:
Oh, I meant ValueError from parsing `float(cpu...)`. A safeguard to remove the assumption that arg `cpu` is validated upstream.

WoosukKwon replied:
I feel it's redundant. In which case do you think we'll need it?

concretevitamin replied:
E.g., in the future somehow the service_catalog function gets called elsewhere.

WoosukKwon replied:
Alright. Added.

concretevitamin commented on sky/clouds/aws.py:
```diff
@@ -350,16 +349,21 @@ def _make(instance_list):
-            return (_make([AWS.get_default_instance_type()]),
-                    fuzzy_candidate_list)
+            # Return a default instance type with the given number of vCPUs.
+            default_instance_type = AWS.get_default_instance_type(
```
For discussion/a thought: This is now dynamically picking a VM type based on `cpu`, which means it's no longer a "default" constant. More like a `service_catalog.get_instance_type_for_cpu()` a la L371.  

Should we remove this interface (Cloud.get_default_instance_type()) then? Wdyt? (I can see arguments for both choices.)

WoosukKwon replied:
I think it cannot be `get_instance_type_for_cpu()` because it limits the search space to the **default** instance family and uses 8 as the **default** number of the vCPUs. I think one can easily understand what it means.

concretevitamin replied:
The naming part sounds good. Sorry, I did not spell out in my comment but was thinking removing one interface from Cloud will make it easier to add new clouds. Also, it's a shallow method (1-line dispatch). How about removing it and just calling `service_catalog.get_default_instance_type()` directly here?

WoosukKwon replied:
Wouldn't it be still useful to explicitly show what the default type is?

concretevitamin replied:
Do you mind clarifying what you meant? 

Currently, the Cloud interface includes a shallow method, which does not show the default type:
```python
    @classmethod
    def get_default_instance_type(cls,
                                  cpu: Optional[str] = None) -> Optional[str]:
        return service_catalog.get_default_instance_type(cpu=cpu, clouds='aws')
```

WoosukKwon replied:
I mean one can easily find which instance type is default for AWS by calling `AWS.get_default_instance_type`, which is good.

concretevitamin replied:
Makes sense.

concretevitamin commented on sky/clouds/gcp.py:
```diff
@@ -382,40 +381,69 @@ def get_feasible_launchable_resources(self, resources):
+            # Return a default instance type with the given number of vCPUs.
+            host_vm_type = GCP.get_default_instance_type(cpu=resources.cpu)
+            if host_vm_type is None:
                 return ([], fuzzy_candidate_list)
```
ditto, return [] ?

WoosukKwon replied:
Fixed.

concretevitamin commented on sky/clouds/service_catalog/aws_catalog.py:
```diff
@@ -19,6 +19,11 @@
 
 logger = sky_logging.init_logger(__name__)
 
+# General-purpose instance with Intel Ice Lake 8375C
```
Add: This is the latest general-purpose instance family as of Jan 2023.

WoosukKwon replied:
Added.

concretevitamin commented on sky/clouds/service_catalog/common.py:
```diff
@@ -270,6 +270,25 @@ def get_vcpus_from_instance_type_impl(
+        return df[df['vCPUs'] == float(cpu)]
+
+
+def get_default_instance_type_impl(df: pd.DataFrame,
+                                   cpu: Optional[str] = None) -> Optional[str]:
+    df = _filter_with_cpu(df, cpu)
+    if df.empty:
+        return None
+    # Sort by the number of vCPUs and then by the price.
+    df = df.sort_values(by=['vCPUs', 'Price'], ascending=True)
+    return df['InstanceType'].iloc[0]
```
See comment in aws.py. This feels more like filtering vs. getting the default?

WoosukKwon replied:
For this function, I agree that it has nothing to do with "default". Renamed to `get_instance_type_for_cpu_impl`.

concretevitamin commented on sky/clouds/service_catalog/aws_catalog.py:
```diff
@@ -94,6 +99,17 @@ def get_vcpus_from_instance_type(instance_type: str) -> Optional[float]:
+def get_default_instance_type(cpu: Optional[str] = None) -> Optional[str]:
+    if cpu is None:
+        cpu = str(_DEFAULT_NUM_VCPUS)
+    # The metal instance is not included in the default instance family.
```
Why do we exclude them?

WoosukKwon replied:
Good point. Fixed.

concretevitamin commented on sky/clouds/service_catalog/aws_catalog.py:
```diff
@@ -94,6 +99,17 @@ def get_vcpus_from_instance_type(instance_type: str) -> Optional[float]:
+    instance_type_prefix = f'{_DEFAULT_INSTANCE_FAMILY}.'
+    instance_type_suffix = 'xlarge'
+    df = _df[_df['InstanceType'].str.startswith(instance_type_prefix) &
+             _df['InstanceType'].str.endswith(instance_type_suffix)]
```
Like above: why not just filter by the prefix + '.'?

WoosukKwon replied:
Good idea. Fixed.

concretevitamin commented on sky/clouds/service_catalog/azure_catalog.py:
```diff
@@ -11,6 +12,11 @@
 
 _df = common.read_catalog('azure/vms.csv')
 
+# General-purpose instance with Intel Ice Lake 8370C
```
Add: This is the latest general-purpose instance family as of Jan 2023.

WoosukKwon replied:
Added.

concretevitamin commented on sky/execution.py:
```diff
@@ -460,6 +460,19 @@ def exec(  # pylint: disable=redefined-builtin
+    else:
+        task = entrypoint
+    task_resources = task.get_resources()
+    if task_resources:
```
Is this testing for None? Looks like we always set https://github.com/skypilot-org/skypilot/blob/9324e524d04a3cb3d4de136467bedf5f35334996/sky/task.py#L156.

WoosukKwon replied:
Good point. I thought `resources` can be an empty set, but it seems such a case cannot exist. Removed `if task_resources`.

concretevitamin commented on sky/execution.py:
```diff
@@ -460,6 +460,19 @@ def exec(  # pylint: disable=redefined-builtin
+        task_resources = list(task_resources)[0]
+        if task_resources.cpu is not None:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(f'Cannot specify CPU resources for sky exec. '
```
```suggestion
                raise ValueError(f'Cannot specify CPU resources for sky.exec(). '
```

WoosukKwon replied:
Actually, that's exactly what the copilot suggested. However, I intentionally changed it into `sky exec` for the sake of  CLI users.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -25,6 +25,12 @@
     'asia-east1',
 ]
 
+# Default instance family for CPU-only VMs.
```
Add: This is the latest general-purpose instance family as of Jan 2023.

WoosukKwon replied:
Added.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -25,6 +25,12 @@
+# Default instance family for CPU-only VMs.
+# General-purpose instance with Intel Ice Lake 8373C or Cascade Lake 6268CL
+# 4 GB RAM per 1 vCPU
+_DEFAULT_INSTANCE_FAMILY = 'n2'
```
```suggestion
_DEFAULT_INSTANCE_FAMILY = 'n2-standard'
```
?

WoosukKwon replied:
I think either makes sense; it depends on how we define "instance family". I just tried to make it consistent with `_DEFAULT_HOST_VM_FAMILY = 'n1'`.

concretevitamin replied:
Does GCP define this term?

If not (or even if it does), I'd prefer to spell it out at the constants' definitions, otherwise the reader is left wondering whether it's standard/highmem/.., and it's only revealed by poking deep into the implementation L177.

WoosukKwon replied:
> Does GCP define this term?

To my understanding, their official terminologies "machine series" (e.g., N1, N2, N2D) and "machine types" (e.g, n1-highmem-8). The "instance family" is our own terminology.

> If not (or even if it does), I'd prefer to spell it out at the constants' definitions, otherwise the reader is left wondering whether it's standard/highmem/.., and it's only revealed by poking deep into the implementation L177.

Alright. Fixed.

concretevitamin reviewed (commented):
Very nice @WoosukKwon. Did a pass and left some comments.

1. Tests: consider spelling out the manual tests. Also, should we add some dryrun unit tests to make sure `sky.optimize()` works for things like `sky.Resources(..., cpu=...)`?

2. Should we warn + ignore `--cpu` in `sky launch -c existing --cpu x`?

WoosukKwon commented on sky/resources.py:
```diff
@@ -375,6 +416,26 @@ def _try_validate_instance_type(self) -> None:
+                            f'{self.instance_type} does not have enough vCPUs. '
+                            f'{self.instance_type} has {cpu} vCPUs, '
+                            f'but {self.cpu} is requested.')
+            elif cpu != self.cpu:
```
Good catch! Fixed.

WoosukKwon commented on sky/resources.py:
```diff
@@ -375,6 +416,26 @@ def _try_validate_instance_type(self) -> None:
+        if self.cpu is None:
+            return
+        if self.instance_type is not None:
+            assert self.cloud is not None
```
Because of `_try_validate_instance_type` which is executed before this method. When the instance type is specified by the user, `Resources` infers the cloud provider from the catalogs and set its `self._cloud`.

concretevitamin replied:
Thanks, please add this as an inline comment to make it clearer for future reading :)

WoosukKwon replied:
Added.

WoosukKwon commented on sky/cli.py:
```diff
@@ -174,6 +174,11 @@ def _interactive_node_cli_command(cli_func):
+    cpu = click.option('--cpu',
+                       default=None,
+                       type=str,
+                       help='Number of vCPUs to use. '
+                       '(e.g., ``--cpu=4`` or ``--cpu=4+``).')
```
Thanks for the suggestion. Fixed.

WoosukKwon commented on sky/cli.py:
```diff
@@ -1090,6 +1104,11 @@ def cli():
+              default=None,
+              type=str,
+              required=False,
+              help='The number of vCPUs to use. Either <number> or <number>+.')
```
Fixed.

WoosukKwon commented on sky/cli.py:
```diff
@@ -2967,6 +2995,11 @@ def spot():
+              default=None,
+              type=str,
+              required=False,
+              help='The number of vCPUs to use. Either <number> or <number>+.')
```
Fixed.

WoosukKwon commented on sky/resources.py:
```diff
@@ -188,6 +195,10 @@ def zone(self):
         return self._instance_type
 
+    @property
+    def cpu(self) -> Optional[str]:
```
Added, and I also emphasized that the Resources in a ResourceHandle must be set to None. PTAL.

WoosukKwon commented:
Thanks @concretevitamin for the quick review. I've addressed most of your reviews. PTAL.

> 1. Tests: consider spelling out the manual tests. Also, should we add some dryrun unit tests to make sure sky.optimize() works for things like sky.Resources(..., cpu=...)?
Will add them soon.

> 2. Should we warn + ignore --cpu in sky launch -c existing --cpu x?
Nice catch! Fixed.

concretevitamin commented on sky/clouds/cloud.py:
```diff
@@ -181,7 +181,13 @@ def get_accelerators_from_instance_type(
     def get_default_instance_type(cls,
-                                  cpu: Optional[str] = None) -> Optional[str]:
+                                  cpus: Optional[str] = None) -> Optional[str]:
+        """Returns the default instance type with the given number of CPUs.
```
Document the two formats as well?

WoosukKwon replied:
Added.

concretevitamin commented on sky/resources.py:
```diff
@@ -41,17 +41,15 @@ class Resources:
-    # 2. Change the __setstate__ method to handle the new fields.
-    # 3. Modify the to_config method to handle the new fields.
+    # If any fields changed, increment the version. For backward compatibility,
+    # modify the __setstate__ method to handle the old version.
```
We do need the step of changing `from_yaml_config()` too?

WoosukKwon replied:
I don't think it's particularly relevant to backward compatibility, do you?

concretevitamin replied:
Makes sense. Was thinking this comment is a guide for "all the places we need to change when we add a new field". For just back compat, it seems `__setstate__` is the only required change. cc @Michaelvll to double check

WoosukKwon replied:
I think it's hard to provide such a comprehensive guideline, because in the worst case almost all the methods would have to be fixed. It's her responsibility to add necessary changes when a new field is added.

concretevitamin commented on sky/resources.py:
```diff
@@ -196,8 +194,15 @@ def instance_type(self):
-    def cpu(self) -> Optional[str]:
-        return self._cpu
+    def cpus(self) -> Optional[str]:
+        """Returns the number of vcpus that each instance must have.
```
```suggestion
        """Returns the number of vCPUs that each instance must have.
```

WoosukKwon replied:
Good catch. Fixed.

concretevitamin commented on sky/resources.py:
```diff
@@ -196,8 +194,15 @@ def instance_type(self):
+        """Returns the number of vcpus that each instance must have.
+
+        For example, cpus='4' means each instance must have exactly 4 vcpus,
+        and cpus='4+' means each instance must have at least 4 vcpus.
```
```suggestion
        and cpus='4+' means each instance must have at least 4 vcpus.
        
```

(styleguide) break paragraphs by blank lines

WoosukKwon replied:
Good catch. Fixed.

concretevitamin commented on sky/execution.py:
```diff
@@ -176,6 +176,15 @@ def _execute(
             cluster_name)
         cluster_exists = existing_handle is not None
+    if cluster_exists:
+        assert len(task.resources) == 1
```
Come to think of it, one can indeed do `Task().set_resources({})` or pass None there. Should we fix this in `set_resources()`?

WoosukKwon replied:
Maybe we can prevent such a case in the `task.set_resources` method. I think it's not relevant to this PR though.

concretevitamin replied:
Yes, please add a TODO?

WoosukKwon replied:
Added.

concretevitamin commented on sky/resources.py:
```diff
@@ -196,8 +194,15 @@ def instance_type(self):
+        For example, cpus='4' means each instance must have exactly 4 vcpus,
+        and cpus='4+' means each instance must have at least 4 vcpus.
+        The cpus field is only used to select the instance type at launch time.
+        Thus, Resources in ResourceHandle will have cpus field set to None.
```
```suggestion
        (Developer note: Thus, Resources in the backend's ResourceHandle will have cpus field set to None.)
```

WoosukKwon replied:
Good catch. Marked the last two sentences as a developer note. PTAL.

concretevitamin reviewed (commented):
Looking good! Mostly minor comments.

concretevitamin reviewed (commented):
Thanks @WoosukKwon. Some final comments & reminder to add some tests to maybe `test_optimizer_dryruns.py`.

concretevitamin commented on tests/test_optimizer_dryruns.py:
```diff
@@ -148,6 +163,41 @@ def test_clouds_not_enabled(monkeypatch):
+    _test_resources_launch(monkeypatch,
+                           sky.AWS(),
+                           instance_type='g4dn.2xlarge',
+                           cpus=8)
```
Maybe add the `8.0` and `'8'` variants for the last resource.

WoosukKwon replied:
Fixed.

concretevitamin reviewed (approved):
Thanks for the awesome contribution @WoosukKwon. A major update to our Resources in a long time. LGTM.

PS: consider some basic smoke tests (e.g., minimal) before merging?

WoosukKwon commented:
@concretevitamin This PR has passed all the smoke tests.

concretevitamin commented:
Missed this before: shall we add the new field to https://skypilot.readthedocs.io/en/latest/reference/yaml-spec.html?

WoosukKwon commented:
Good catch. Added.

dongreenberg commented:
Woo!!! Congrats and thanks gang!! Can't wait to test it out!
